client/Windows/wf_defaults.c

```diff
@@ -0,0 +1,112 @@
+#include <Windows.h>
+#include <wincred.h>
+#include <stdio.h>
+#include <malloc.h>
+
+#pragma warning(disable : 4706 4094) // assignment within conditional expression
+
+#define FreeRDP_ServerHostname (20)
+#define FreeRDP_Username (21)
+#define FreeRDP_Password (22)
+#define FreeRDP_GatewayHostname (1986)
+#define FreeRDP_GatewayUsername (1987)
+#define FreeRDP_GatewayPassword (1988)
+
+typedef struct rdp_settings rdpSettings;
+
+EXTERN_C_START
+
+const char* freerdp_settings_get_string(const rdpSettings* settings, size_t id);
+BOOL freerdp_settings_set_string(rdpSettings* settings, size_t id, const char* param);
+
+EXTERN_C_END
+
+PCWSTR ValidateString(const BYTE* pb, ULONG cb)
+{
+	return pb &&                                       // present ?
+	               cb &&                               // size not 0 ?
+	               !(cb & (sizeof(WCHAR) - 1)) &&      // cb == n * sizeof(WCHAR) ?
+	               !*(WCHAR*)(pb + cb - sizeof(WCHAR)) // 0 terminated ?
+	           ? (PCWSTR)pb
+	           : 0;
+}
+
+void AddDefaultSettings_I(rdpSettings* settings, size_t idHostname, size_t idUsername,
+                          size_t idPassword)
+{
+	PCSTR ServerHostname = freerdp_settings_get_string(settings, idHostname);
+
+	if (!ServerHostname)
+	{
+		return;
+	}
+
+	BOOL bExistUserName = 0 != freerdp_settings_get_string(settings, idUsername);
+	BOOL bExistPassword = 0 != freerdp_settings_get_string(settings, idPassword);
+
+	if (bExistUserName && bExistPassword)
+	{
+		return;
+	}
+
+	PCREDENTIALA Credential;
+	PCWSTR lpWideCharStr;
+
+	PSTR Password;
+	PSTR UserName;
+
+	PSTR TargetName = 0;
+	int len = 0;
+
+	while (0 < (len = _vsnprintf(TargetName, len, "TERMSRV/%s", (va_list)&ServerHostname)))
```

Comment:
shouldn´t this simply be `_vsnprintf(TargetName, len, "TERMSRV/%s", ServerHostname)` ?

Reply:
Yeah I had to re-read it twice to see what was checked

Reply:
hmm, doesn't work or maybe I'm doing something wrong? how would you rewrite that code without the while?

Reply:
I should mention that I'm not really a C dev. Any help would be appreciated.

Reply:
removing the while I end up with something like this:

```C
	PSTR TargetName = 0;
	int len = _vsnprintf(TargetName, 0, "TERMSRV/%s", (va_list)&ServerHostname);
	TargetName = (PSTR)malloc(++len);
	_vsnprintf(TargetName, len, "TERMSRV/%s", (va_list)&ServerHostname);

```

is this the preferred way? (btw, I failed using _snprintf)